net/sockstats/sockstats_tsgo.go

```diff
@@ -276,3 +283,71 @@ func isLikelyCellularInterface(ifName string) bool {
 		strings.HasPrefix(ifName, "ww") || // systemd naming scheme for WWAN
 		strings.HasPrefix(ifName, "pdp") // iOS
 }
+
+// radio very roughly approximates the power state transitions of a cellular radio.
+// Time spent in the high and low power states is loggged as client metrics.
+var radio = struct {
+	mu       sync.Mutex
+	state    radioState // current radio power state
+	wakeTime time.Time  // time radio was moved into high power state
+
+	timerHigh *time.Timer // timer for when radio transitions out of high power state
+	timerLow  *time.Timer // timer for when radio transitions out of low power state
+
+	highMetric, lowMetric *clientmetric.Metric // metrics tracking time spent in power states
+}{
+	highMetric: clientmetric.NewCounter("sockstats_radio_high_duration_ms"),
+	lowMetric:  clientmetric.NewCounter("sockstats_radio_low_duration_ms"),
+}
+
+// radioActivity should be called whenever network activity occurs on a cellular network interface.
+func radioActivity() {
+	setRadioState(radioHighPower)
+}
+
+type radioState int
+
+const (
+	radioOff radioState = iota
+	radioLowPower
+	radioHighPower
+)
+
+// Timings for radio power state transitions taken from
+// https://developer.android.com/training/connectivity/network-access-optimization#radio-state
+// Even though that documents a typical 3G radio and newer radios are much more efficient,
+// it provides worst-case timings to use for analysis.
+const (
+	radioHighIdle = 5 * time.Second  // time radio idles in high power state before transitioning to low
+	radioLowIdle  = 12 * time.Second // time radio idles in low power state before transitioning to off
+)
+
+func setRadioState(newState radioState) {
+	radio.mu.Lock()
+	defer radio.mu.Unlock()
+
+	switch newState {
+	case radioHighPower:
+		if radio.state != newState {
+			radio.wakeTime = time.Now()
+			if radio.timerLow != nil {
+				radio.timerLow.Stop()
```

Comment:
I think you'd want to increment `lowMetric` in this case too, for cases where the radio gets turned on again without making it all the way to an off state.

Reply:
ah yeah, you're right.  Went too far with trying to simplify this.